picking_good_slis.md

```diff
@@ -0,0 +1,92 @@
+# Picking Good SLIs
+
+## SRE's Philosophy on Monitoring
+
+A core goal of SRE is to ensure the availability of a service. "Availability"
+here means much more than simply "is my application running." Effectively
+monitoring a modern cloud application requires insight into the application's
+response time, error rate, traffic level, and more.
+
+Furthermore, the scope of monitoring should not be limited solely for use in
+deciding when to page an operations team. Metrics should be closely tied to
+business objectives, informing decisions on what features to prioritize and
+when to focus on adding new features versus improving reliability.
+
+To make this vision a reality, SRE teams rely on two key concepts:
+  * Service Level Indicators (SLIs) - A quantifiable metric that measures a
+    specific aspect of a service's availability.
+  * Service Level Objectives (SLOs) - A measurable, specific goal for a
+    given service's availability over time, based on one or more SLIs.
+
+As a starting point on towards realizing this goal, this document aims to
+provide guidance on selecting the right SLIs for your service.
+
+## Selecting SLIs
+
+For SRE teams looking to start monitoring their services, start by identifying
+SLIs to measure. Focus on measuring the basics to start, can refine them over
+time as you gain more experience operating the service.
+
+A good starting point is to attempt to implement an SLI for each of the four
+"Golden Signals".
+
+## The Four Golden Signals of Monitoring
+
+Google defines the four golden signals well in their [SRE Book][1]: "The four
+golden signals of monitoring are latency, traffic, errors, and saturation. If
+you can only measure four metrics of your user-facing system, focus on these
+four."
+
+We recommend implementing at least one SLI for each of these signals.
+
+### Latency
+
+Latency measures how long it takes your service to handle a request. Examples
+of how to measure this might be:
+  * How long does it take for a page of your browser-based application to load
+  * How long does a database query take
+  * How long does it take to provision some user-requested environment or
+    resource
+
+You will likely want to distinguish between the latency of successful and
+failed requests in your measurements and ultimatley create separate SLOs for
+both.
```

Comment:
I disagree that there should be a latency SLO for failed requests. Failed requests should be handled by an availability SLO.

Reply:
@davidffrench Time to fail is a fair thing to be concerned about (though I'm not sure it's a beginner's SLO). If a request is "doomed" (it's going to fail at some point in a request chain) then it makes a lot of sense to figure that out as soon as possible. Identifying "doomed" requests and notifying the requestor means the feedback cycle is shortened, and possibly less resource consumption on the service side.

In real terms: If a customer is asking for a cluster to be installed, but there's a fatal fault with their cloud account, we should have pre-flight checks in place so a potentially long install process doesn't abort after some long period of time where we keep the customer waiting.

Reply:
I copied this comment into the Google doc for further discussion.